src/opf/ptdf.jl

```diff
@@ -0,0 +1,112 @@
+using SparseArrays
+using PowerModels
+using LinearAlgebra
+
+abstract type AbstractPTDF end
+
+
+struct LazyPTDF{TF} <: AbstractPTDF
+    N::Int  # number of buses
+    E::Int  # number of branches
+    islack::Int  # Index of slack bus
+
+    A::SparseMatrixCSC{Float64,Int}  # incidence matrix
+    b::Vector{Float64}  # branch susceptances
+    BA::SparseMatrixCSC{Float64,Int}  # B*A
+    AtBA::SparseMatrixCSC{Float64,Int}  # AᵀBA
+
+    F::LDLt # Factorization of AᵀBA. Must be able to solve linear systems with F \ p
+
+    # TODO: cache
+end
+
+function LazyPTDF(data::OPFData)
+    N, E, A, b, ref_idx = data.N, data.E, data.A, data.b, data.ref_bus
+
+    all(data.branch_status .== 1) || error("LazyPTDF does not support disabled branches.")
+
+    B = Diagonal(b)
+    BA = B * A
+    S = AtBA = A' * BA
+
+    S[ref_idx, :] .= 0.0
+    S[:, ref_idx] .= 0.0
+    S[ref_idx, ref_idx] = 1.0
+
+    F = ldlt(S)
+
+    return LazyPTDF(N, E, ref_idx, A, b, BA, AtBA, F)
+end
+
+"""
+    compute_flow!(pf, pg, Φ::LazyPTDF)
+
+Compute power flow `pf = Φ*pg` lazily, without forming the PTDF matrix.
+
+Namely, `pf` is computed as `pf = BA * (F \\ pg)`, where `F` is a factorization
+    of (-AᵀBA), e.g., a cholesky / LDLᵀ / LU factorization.
```

Comment:
Update docstring to reflect sign convention
```suggestion
Namely, `pf` is computed as `pf = BA * (F \\ pg)`, where `F` is an LDLᵀ factorization of `AᵀBA`.
```